lib/provisioner/worker.rb

```diff
@@ -0,0 +1,322 @@
+
+    # Instantiate and run an instance of given plugin for a task
+    def _run_plugin(clazz, env, cwd, task)
+      clusterId = task['clusterId']
+      hostname = task['config']['hostname']
+      provider = task['config']['provider']['description']
+      imagetype = task['config']['imagetype']
+      hardware = task['config']['hardwaretype']
+      taskName = task['taskName'].downcase
+      log.info "Creating node #{hostname} on #{provider} for #{clusterId} using #{imagetype} on #{hardware}" if taskName == 'create'
+
+      object = clazz.new(env, task)
+      FileUtils.mkdir_p(cwd)
+      Dir.chdir(cwd) do
+        result = object.runTask
+        log.info "#{clusterId} on #{hostname} could not be deleted: #{result['message']}" if taskName == 'delete' && result['status'] != 0
+        result
+      end
+    end
+
+    def delegate_task(task, pluginmanager)
+      providerName = nil # rubocop:disable UselessAssignment
+      automatorName = nil # rubocop:disable UselessAssignment
+      clazz = nil # rubocop:disable UselessAssignment
+      object = nil
+      result = nil
+      classes = nil
+      task_id = task['taskId']
+
+      log.debug "Processing task with id #{task_id} ..."
+
+      taskName = task['taskName'].downcase
+      # depending on task, these may be nil
+      # automator take pecedence as presence indicates a 'software' task
+      providerName = task['config']['provider']['providertype'] rescue nil
+      automatorName = task['config']['service']['action']['type'] rescue nil
+
+      case taskName.downcase
+      when 'create', 'confirm', 'delete'
+        clazz = Object.const_get(pluginmanager.getHandlerActionObjectForProvider(providerName))
+        cwd = File.join(@config.get(PROVISIONER_WORK_DIR), @tenant, 'providertypes', providerName)
+        result = _run_plugin(clazz, @plugin_env, cwd, task)
+      when 'install', 'configure', 'initialize', 'start', 'stop', 'remove'
+        clazz = Object.const_get(pluginmanager.getHandlerActionObjectForAutomator(automatorName))
+        cwd = File.join(@config.get(PROVISIONER_WORK_DIR), @tenant, 'automatortypes', automatorName)
+        result = _run_plugin(clazz, @plugin_env, cwd, task)
+      when 'bootstrap'
+        combinedresult = {}
+        classes = []
+        if task['config'].key? 'automators' and !task['config']['automators'].empty?
+          # server must specify which bootstrap handlers need to run
+          log.debug "Task #{task_id} running specified bootstrap handlers: #{task['config']['automators']}"
+          task['config']['automators'].each do |automator|
+            clazz = Object.const_get(pluginmanager.getHandlerActionObjectForAutomator(automator))
+            cwd = File.join(@config.get(PROVISIONER_WORK_DIR), @tenant, 'automatortypes', automator)
+            result = _run_plugin(clazz, @plugin_env, cwd, task)
+            combinedresult.merge!(result)
+          end
+        else
+          log.warn 'No automators specified to bootstrap'
```

Comment:
Should this be a failure? If there's no automators to be run, there's no reason to bootstrap. If we're hitting this case, something is wrong, anyway.